File: backend/src/data/store.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const DB_FILE = path.join(__dirname, 'transactions.json');

let transactions = [];

const seedData = () => {
  const generated = [];
  const start = new Date('2025-09-01').getTime();
  const end = new Date('2026-03-25').getTime();
  const expCats = ['Rent', 'Salaries', 'Inventory', 'Marketing', 'Utilities', 'Software'];
  const incCats = ['Sales', 'Services', 'Consulting'];

  for (let i = 0; i < 115; i++) {
    const time = start + Math.random() * (end - start);
    const dateStr = new Date(time).toISOString().split('T')[0];
    const isIncome = Math.random() > 0.65;
    const type = isIncome ? 'income' : 'expense';
    const amount = Math.floor(Math.random() * 80000) + 1500;
    
    let category = '';
    let description = '';
    
    if (isIncome) {
      category = incCats[Math.floor(Math.random() * incCats.length)];
      description = `Invoice #${1000 + Math.floor(Math.random() * 9000)} Payment`;
    } else {
      category = expCats[Math.floor(Math.random() * expCats.length)];
      description = `Expense - ${category} ${Math.floor(Math.random() * 100)}`;
    }
    
    generated.push({
      id: Date.now().toString() + i + Math.random().toString().slice(2,8),
      date: dateStr,
      description,
      amount,
      type,
      category
    });
  }
  
  generated.sort((a, b) => new Date(a.date) - new Date(b.date));
  return generated;
};

const loadData = () => {
  if (fs.existsSync(DB_FILE)) {
    try {
      const data = fs.readFileSync(DB_FILE, 'utf8');
      const parsed = JSON.parse(data);
      if (parsed.length > 0) {
        transactions = parsed;
        return;
      }
    } catch (e) {
      console.error('Error reading DB:', e);
    }
  }
  transactions = seedData();
  saveData();
};

const saveData = () => {
  try {
    fs.writeFileSync(DB_FILE, JSON.stringify(transactions, null, 2), 'utf8');
  } catch (e) {
    console.error('Error saving DB:', e);
  }
};

loadData();

export const getTransactions = () => transactions;

export const clearTransactions = () => { 
  transactions = []; 
  saveData(); 
};

export const addTransaction = (txn) => {
  const newTxn = { ...txn, id: Date.now().toString() + Math.random().toString().slice(2,8) };
  transactions.push(newTxn);
  saveData();
  return newTxn;
};

export const addTransactions = (txns) => {
  const newTxns = txns.map(t => ({ ...t, id: Date.now().toString() + Math.random().toString().slice(2,8) }));
  transactions = [...transactions, ...newTxns];
  saveData();
  return newTxns;
};

export const deleteTransaction = (id) => {
  const idx = transactions.findIndex(t => t.id === id);
  if (idx === -1) return false;
  transactions.splice(idx, 1);
  saveData();
  return true;
};

export const updateTransaction = (id, updates) => {
  const idx = transactions.findIndex(t => t.id === id);
  if (idx === -1) return null;
  transactions[idx] = { ...transactions[idx], ...updates };
  saveData();
  return transactions[idx];
};

File: backend/store.py
import json
import os
import random
from datetime import datetime

# Build path to the JSON file
# Use /tmp for writable storage on Render (ephemeral but works within session)
DB_FILE = os.path.join("/tmp", "vantge_transactions.json")

# Make sure directory exists
os.makedirs(os.path.dirname(DB_FILE), exist_ok=True)

transactions = []

def seed_data():
    generated = []
    start = datetime(2025, 9, 1).timestamp()
    end = datetime(2026, 3, 25).timestamp()
    exp_cats = ['Rent', 'Salaries', 'Inventory', 'Marketing', 'Utilities', 'Software']
    inc_cats = ['Sales', 'Services', 'Consulting']

    for i in range(115):
        time_ts = start + random.random() * (end - start)
        date_str = datetime.fromtimestamp(time_ts).strftime('%Y-%m-%d')
        is_income = random.random() > 0.65
        t_type = 'income' if is_income else 'expense'
        amount = random.randint(1500, 81500)
        
        if is_income:
            category = random.choice(inc_cats)
            description = f"Invoice #{random.randint(1000, 9999)} Payment"
        else:
            category = random.choice(exp_cats)
            description = f"Expense - {category} {random.randint(0, 99)}"
            
        generated.append({
            "id": str(int(datetime.now().timestamp() * 1000)) + str(i) + str(random.random())[2:8],
            "date": date_str,
            "description": description,
            "amount": amount,
            "type": t_type,
            "category": category
        })
        
    generated.sort(key=lambda x: x['date'])
    return generated

def load_data():
    global transactions
    if os.path.exists(DB_FILE):
        try:
            with open(DB_FILE, 'r', encoding='utf-8') as f:
                data = json.load(f)
                if len(data) > 0:
                    transactions = data
                    return
        except Exception as e:
            print("Error reading DB:", e)
    
    transactions = seed_data()
    save_data()

def save_data():
    global transactions
    try:
        with open(DB_FILE, 'w', encoding='utf-8') as f:
            json.dump(transactions, f, indent=2)
    except Exception as e:
        print("Error saving DB:", e)

# Initial load
load_data()

def get_transactions():
    return transactions

def clear_transactions():
    global transactions
    transactions = []
    save_data()

def add_transaction(txn: dict):
    global transactions
    new_txn = txn.copy()
    new_txn["id"] = str(int(datetime.now().timestamp() * 1000)) + str(random.random())[2:8]
    transactions.append(new_txn)
    save_data()
    return new_txn

def add_transactions(txns: list):
    global transactions
    new_txns = []
    for t in txns:
        new_txn = t.copy()
        new_txn["id"] = str(int(datetime.now().timestamp() * 1000)) + str(random.random())[2:8]
        new_txns.append(new_txn)
    transactions.extend(new_txns)
    save_data()
    return new_txns

def delete_transaction(tid: str):
    global transactions
    for i, t in enumerate(transactions):
        if t["id"] == tid:
            del transactions[i]
            save_data()
            return True
    return False

def update_transaction(tid: str, updates: dict):
    global transactions
    for i, t in enumerate(transactions):
        if t["id"] == tid:
            transactions[i].update(updates)
            save_data()
            return transactions[i]
    return None
